File: backend/app/models/candidate.py
"""
Candidate entity matching Java Candidate.java.
One-to-one relationship with User using @MapsId pattern.
"""
from datetime import datetime
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey
from sqlalchemy.orm import relationship

from app.db.base import Base


class Candidate(Base):
    """
    Candidate-specific profile data, one row per candidate User.
    
    The 1:1 link to User uses the User's own primary key as the foreign key
    (matching Java @MapsId pattern): candidates.id is both the PK and a FK to users.id.
    """
    __tablename__ = "candidates"
    
    id = Column(Integer, ForeignKey('users.id', ondelete='CASCADE'), primary_key=True)
    phone = Column(String(20), nullable=True)
    location = Column(String(150), nullable=True)
    bio = Column(String(2000), nullable=True)
    github_url = Column(String(255), nullable=True)
    linkedin_url = Column(String(255), nullable=True)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    user = relationship("User", back_populates="candidate")
    
    # Skills managed via CandidateSkill ORM model
    _skills = relationship(
        "CandidateSkill",
        cascade="all, delete-orphan",
        lazy="joined"
    )
    
    @property
    def skills(self) -> list[str]:
        """Get skills as a list of strings."""
        return [skill.skill for skill in self._skills]
    
    @skills.setter
    def skills(self, value: list[str]):
        """Set skills from a list of strings."""
        # This will be handled in the repository layer
        pass


# Helper model for skills management
class CandidateSkill(Base):
    """Helper model to manage candidate skills through ORM."""
    __tablename__ = "candidate_skills"
    
    candidate_id = Column(Integer, ForeignKey('candidates.id', ondelete='CASCADE'), primary_key=True)
    skill = Column(String(100), primary_key=True)

File: backend/src/main/java/com/codeforge/entity/Candidate.java
package com.codeforge.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

/**
 * Candidate profile entity (1:1 with User, role=CANDIDATE)
 */
@Entity
@Table(name = "candidates")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(exclude = {"user"})
@ToString(exclude = {"user"})
public class Candidate {
    
    @Id
    private Long id;
    
    @MapsId
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id")
    private User user;
    
    @Column(length = 20)
    private String phone;
    
    @Column(length = 255)
    private String location;
    
    @Column(columnDefinition = "TEXT")
    private String bio;
    
    @Column(length = 500)
    private String githubUrl;
    
    @Column(length = 500)
    private String linkedinUrl;
    
    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @UpdateTimestamp
    @Column(nullable = false)
    private LocalDateTime updatedAt;
}
